Improve diagnostic output of `non_local_definitions` lint

Urgau opened the pull request:
This PR improves (or at least tries to improve) the diagnostic output of `non_local_definitions` lint, by simplifying the wording, by adding a "sort of" explanation of bounds interaction that leak the impl...

This PR is best reviewed commit by commit and is voluntarily made a bit vague as to have a starting point to improve on.

Related to https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/non_local_defs.20wording.20improvements

Fixes https://github.com/rust-lang/rust/issues/125068
Fixes https://github.com/rust-lang/rust/issues/124396
cc @workingjubilee
r? @estebank

Urgau committed:
non_local_defs: move out from `#[derive(LintDiagnostic)]` to manual impl

Urgau committed:
non_local_defs: switch to more friendly primary message

Urgau committed:
non_local_defs: improve some notes around trait, bounds, consts

 - Restrict const-anon exception diag to relevant places
 - Invoke bounds (and type-inference) in non_local_defs
 - Specialize diagnostic for impl without Trait

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9069028086/job/24917695607) [(plain)](https://github.com/rust-lang/rust/commit/0c6926510dfe6b68d35943c13ba97ac2916f4887/checks/24917695607/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#17 exporting to docker image format
#17 sending tarball 29.5s done
#17 DONE 34.6s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

---- [ui] tests/ui/proc-macro/nested-macro-rules.rs stdout ----
diff of stderr:

- warning: non-local `macro_rules!` definition, they should be avoided as they go against expectation
+ warning: non-local `macro_rules!` definition, `#[macro_export]` macro should be written at top level module
2   --> $DIR/auxiliary/nested-macro-rules.rs:7:9
4 LL |   macro_rules! outer_macro {


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/nested-macro-rules/nested-macro-rules.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args proc-macro/nested-macro-rules.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/proc-macro/nested-macro-rules.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/nested-macro-rules/a" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/nested-macro-rules/auxiliary" "-Z" "span-debug" "-Z" "macro-backtrace" "--edition=2018"
--- stdout -------------------------------
PRINT-BANG INPUT (DISPLAY): FirstStruct
PRINT-BANG INPUT (DEBUG): TokenStream [
    Ident {
        ident: "FirstStruct",
        span: /checkout/tests/ui/proc-macro/auxiliary/nested-macro-rules.rs:16:14: 16:25 (#6),
]
]
PRINT-ATTR INPUT (DISPLAY): struct FirstAttrStruct {}
PRINT-ATTR INPUT (DEBUG): TokenStream [
        ident: "struct",
        ident: "struct",
        span: /checkout/tests/ui/proc-macro/auxiliary/nested-macro-rules.rs:10:32: 10:38 (#5),
    Ident {
    Ident {
        ident: "FirstAttrStruct",
        span: /checkout/tests/ui/proc-macro/auxiliary/nested-macro-rules.rs:16:27: 16:42 (#6),
    Group {
        delimiter: Brace,
        stream: TokenStream [],
        stream: TokenStream [],
        span: /checkout/tests/ui/proc-macro/auxiliary/nested-macro-rules.rs:10:57: 10:59 (#5),
]
PRINT-BANG INPUT (DISPLAY): SecondStruct
PRINT-BANG INPUT (DISPLAY): SecondStruct
PRINT-BANG INPUT (DEBUG): TokenStream [
        ident: "SecondStruct",
        ident: "SecondStruct",
        span: /checkout/tests/ui/proc-macro/nested-macro-rules.rs:23:38: 23:50 (#15),
]
]
PRINT-ATTR INPUT (DISPLAY): struct SecondAttrStruct {}
PRINT-ATTR INPUT (DEBUG): TokenStream [
        ident: "struct",
        ident: "struct",
        span: /checkout/tests/ui/proc-macro/auxiliary/nested-macro-rules.rs:10:32: 10:38 (#14),
    Ident {
        ident: "SecondAttrStruct",
        ident: "SecondAttrStruct",
        span: /checkout/tests/ui/proc-macro/nested-macro-rules.rs:23:52: 23:68 (#15),
    Group {
        delimiter: Brace,
        stream: TokenStream [],
        stream: TokenStream [],
        span: /checkout/tests/ui/proc-macro/auxiliary/nested-macro-rules.rs:10:57: 10:59 (#14),
]
------------------------------------------
--- stderr -------------------------------
--- stderr -------------------------------
warning: non-local `macro_rules!` definition, `#[macro_export]` macro should be written at top level module
##[warning]  --> /checkout/tests/ui/proc-macro/auxiliary/nested-macro-rules.rs:7:9
LL |   macro_rules! outer_macro {
   |   ------------------------ in this expansion of `nested_macro_rules::outer_macro!`
...
LL | /         macro_rules! inner_macro {
LL | /         macro_rules! inner_macro {
LL | |             ($bang_macro:ident, $attr_macro:ident) => {
LL | |                 $bang_macro!($name);
LL | |                 #[$attr_macro] struct $attr_struct_name {}
LL | |         }
   | |_________^
   |
  ::: /checkout/tests/ui/proc-macro/nested-macro-rules.rs:23:5
  ::: /checkout/tests/ui/proc-macro/nested-macro-rules.rs:23:5
   |
LL |       nested_macro_rules::outer_macro!(SecondStruct, SecondAttrStruct);
   |
   |
   = help: remove the `#[macro_export]` or move this `macro_rules!` outside the of the current function `main`
   = note: a `macro_rules!` definition is non-local if it is nested inside an item and has a `#[macro_export]` attribute
   = note: this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
note: the lint level is defined here
  --> /checkout/tests/ui/proc-macro/nested-macro-rules.rs:8:9
   |
```

</details>

workingjubilee commented on compiler/rustc_lint/messages.ftl:
```diff
@@ -440,17 +440,19 @@ lint_non_local_definitions_cargo_update = the {$macro_kind} `{$macro_name}` may
-    .non_local = an `impl` definition is non-local if it is nested inside an item and may impact type checking outside of that item. This can be the case if neither the trait or the self type are at the same nesting level as the `impl`
-    .exception = one exception to the rule are anon-const (`const _: () = {"{"} ... {"}"}`) at top-level module and anon-const at the same nesting as the trait or type
+    .without_trait = methods and assoc const are still usable outside the current expression, only `impl Local` and `impl dyn Local` are local and only if the `Local` type is at the same nesting as the `impl` block
+    .with_trait = an `impl` definition is non-local if it is nested inside an item and may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
```
hmm so I think we could start more clearly like this:
```
    .with_trait = an `impl` is always non-local
```
with more to it, but "if" is not the next word.

my reasoning is: the impl is non-local by definition, right? it "always is", even if nested. *however*, this only *matters* (affects code in a visible way) if at least one of (list of other considerations we've already discussed to death) applies. so we're choosing to complain about the non-locality in certain cases (roughly "when we believe it matters").

workingjubilee replied:
we might want to try a "yes, *and*" approach like:
```suggestion
    .with_trait = an `impl` is always non-local and may allow access by type inference outside this nesting item if a type and trait is visible
```
am I saying that right? "nesting item" is what I'm saying to try to describe "the thing that the nest*ed* item is *nesting in*". I'm hoping it is somewhat clear from context in actual experience.

Urgau replied:
I like the idea of starting with "is always non-local", but the suggested text implies that the user is trying to "hide" the type, but the note was trying to be more general to also cover cases where neither are being hidden,
```rust
trait Trait {}
struct Type;

fn main() {
    impl Trait for type {}
}
```

what about?
```suggestion
    .with_trait = an `impl` is always non-local, and even when nested inside an item it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
```

workingjubilee replied:
sure!

Urgau replied:
Done in https://github.com/rust-lang/rust/pull/125089/commits/b4024cccaf0ab5a0dbdc88753b618f652bcbda7e.

workingjubilee commented on compiler/rustc_lint/src/non_local_def.rs:
```diff
@@ -215,6 +214,9 @@ impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
                     None
                 };
 
+                let const_anon = matches!(parent_def_kind, DefKind::Const | DefKind::Static { .. })
```
neat!

workingjubilee reviewed (commented):
feel free to "nah actually" I am just trying to throw a couple wording ideas at the wall.

Urgau commented on compiler/rustc_lint/messages.ftl:
```diff
@@ -440,17 +440,19 @@ lint_non_local_definitions_cargo_update = the {$macro_kind} `{$macro_name}` may
         }
-    .non_local = an `impl` definition is non-local if it is nested inside an item and may impact type checking outside of that item. This can be the case if neither the trait or the self type are at the same nesting level as the `impl`
-    .exception = one exception to the rule are anon-const (`const _: () = {"{"} ... {"}"}`) at top-level module and anon-const at the same nesting as the trait or type
+    .without_trait = methods and assoc const are still usable outside the current expression, only `impl Local` and `impl dyn Local` are local and only if the `Local` type is at the same nesting as the `impl` block
```
I think so, but it's not easy and I rather prefer if we do it as a follow-up pr

Urgau committed:
non_local_defs: use labels to indicate what may need to be moved

Urgau committed:
non_local_defs: use span of the impl def and not the impl block

Urgau commented on compiler/rustc_lint/messages.ftl:
```diff
@@ -440,17 +440,19 @@ lint_non_local_definitions_cargo_update = the {$macro_kind} `{$macro_name}` may
 lint_non_local_definitions_deprecation = this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
 
-lint_non_local_definitions_impl = non-local `impl` definition, they should be avoided as they go against expectation
+lint_non_local_definitions_impl = non-local `impl` definition, `impl` blocks should be written at the same level as their item
```
I think it more clearly communicates the intend, at least better than the old one, which was too technical.

Urgau commented:
I pushed several more commits that tries to improve the clarity the message were are trying to pass, mainly by moving the "move help message" to a separate span with some labels.

<details>

```text
warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
  --> $DIR/consts.rs:43:5
   |
LL |     impl Test {
   |     ^^^^^^^^^
   |
   = note: methods and assoc const are still usable outside the current expression, only `impl Local` and `impl dyn Local` are local and only if the `Local` type is at the same nesting as the `impl` block
help: move this `impl` block and outside the of the current function `main`
  --> $DIR/consts.rs:43:5
   |
LL |       impl Test {
   |       ^    ---- may need to be moved as well
   |  _____|
   | |
LL | |
LL | |         fn foo() {}
LL | |     }
   | |_____^
   = note: this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
```

</details>

workingjubilee commented on compiler/rustc_lint/messages.ftl:
```diff
@@ -440,17 +440,19 @@ lint_non_local_definitions_cargo_update = the {$macro_kind} `{$macro_name}` may
+lint_non_local_definitions_impl = non-local `impl` definition, `impl` blocks should be written at the same level as their item
     .help =
-        move this `impl` block outside the of the current {$body_kind_descr} {$depth ->
+        move this `impl` block and outside the of the current {$body_kind_descr} {$depth ->
```
typo?
```suggestion
        move this `impl` block outside the of the current {$body_kind_descr} {$depth ->
```

Urgau replied:
yes, my bad; fixed

workingjubilee commented on compiler/rustc_lint/messages.ftl:
```diff
@@ -440,17 +440,19 @@ lint_non_local_definitions_cargo_update = the {$macro_kind} `{$macro_name}` may
         }
-    .non_local = an `impl` definition is non-local if it is nested inside an item and may impact type checking outside of that item. This can be the case if neither the trait or the self type are at the same nesting level as the `impl`
-    .exception = one exception to the rule are anon-const (`const _: () = {"{"} ... {"}"}`) at top-level module and anon-const at the same nesting as the trait or type
+    .without_trait = methods and assoc const are still usable outside the current expression, only `impl Local` and `impl dyn Local` are local and only if the `Local` type is at the same nesting as the `impl` block
```
I think this might be clearer? slightly reduces [semantic satiation](https://en.wikipedia.org/wiki/Semantic_satiation) at least.
```suggestion
    .without_trait = methods and assoc const are still usable outside the current expression, `impl Local` and `impl dyn Local` are only private if the `Local` type is nested in the same item as the `impl`
```

Urgau replied:
unfortunately this also changes the meaning significantly

what this note is trying to say is that only `impl Local` and `impl dyn Local` can ever be private, and of those two, to be private, `Local` must be at the same nesting as the `impl` def

workingjubilee replied:
hmm so closer to this?
```suggestion
    .without_trait = methods and assoc const are still usable outside the current expression, only `impl Local` and `impl dyn Local` can be private if the type is nested in the same item as the `impl`
```

workingjubilee replied:
( it's not required to find a different wording I'm just trying to figure things out. )

Urgau replied:
idk, feels a bit weird, at least when I try to read it with my non-native english eyes

maybe this?
```suggestion
    .without_trait = methods and associated constants are still usable outside the current expression, only `impl Local` and `impl dyn Local` can ever be private, and only if the type is nested in the same item as the `impl`
```

workingjubilee replied:
Ooh, that seems good.

Urgau replied:
Done.

workingjubilee commented on compiler/rustc_lint/messages.ftl:
```diff
@@ -440,17 +440,19 @@ lint_non_local_definitions_cargo_update = the {$macro_kind} `{$macro_name}` may
+    .without_trait = methods and assoc const are still usable outside the current expression, only `impl Local` and `impl dyn Local` are local and only if the `Local` type is at the same nesting as the `impl` block
+    .with_trait = an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
+    .bounds = `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
+    .exception = anon-const (`const _: () = {"{"} ... {"}"}`) at top-level module and anon-const at the same nesting as the trait or type are consider to be transparent regarding the nesting level
```
what does "transparent regarding the nesting level" mean?

Urgau replied:
it means that putting a non-local `impl` def inside a `const _` will make it "global"
```rust
trait Trait {}
struct Test;

const _: () = {
    impl Trait for Test {} // for the lint Trait, Test and the `impl` are at the same nesting
                           // even through they are clearly not
};
```

workingjubilee replied:
Interesting. I think it might want to say something more like
```suggestion
    .exception = items in an anon const (`const _: () = {"{"} ... {"}"}`) are treated as in the same scope as the anon const's declaration
```

workingjubilee replied:
Mostly "transparent" seems slightly... opaque, ironically, when we want to say "this level of nesting is ignored" or "these two scopes are linted as if the same scope"

Urgau replied:
Simpler, I like it. Updated.

workingjubilee commented on compiler/rustc_lint/messages.ftl:
```diff
@@ -440,17 +440,19 @@ lint_non_local_definitions_cargo_update = the {$macro_kind} `{$macro_name}` may
     .const_anon = use a const-anon item to suppress this lint
 
-lint_non_local_definitions_macro_rules = non-local `macro_rules!` definition, they should be avoided as they go against expectation
+lint_non_local_definitions_macro_rules = non-local `macro_rules!` definition, `#[macro_export]` macro should be written at top level module
```
maybe this?
```suggestion
lint_non_local_definitions_macro_rules = a macro made available by `#[macro_export]` should be written at the top level of the module
```

Urgau replied:
this suggestion feels like it's only explaining the issue and not telling what the issue is, I think we should keep the first part ("non-local `macro_rules!` definition"), we can improve the rest

workingjubilee replied:
suresure.

Urgau committed:
non_local_defs: improve exception note for impl and macro_rules!

 - Remove wrong exception text for non-local macro_rules!
 - Simplify anonymous const exception note

Urgau committed:
non_local_defs: be more precise about what needs to be moved

Urgau committed:
non_local_defs: improve main without a trait note

Urgau committed:
non_local_defs: suggest removing leading ref/ptr to make the impl local

Urgau committed:
non_local_defs: point to Self and Trait to give more context

Urgau commented:
I continued the improved the diagnostic output and I added to labels that points to Self type and Trait to better put emphasis are they are not non-local.

With the example in https://github.com/rust-lang/rust/issues/125068#issue-2292387556, it now gives:
```text
warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
 --> a.rs:6:5
  |
6 |     impl Trait for Option<Local> {}
  |     ^^^^^-----^^^^^------^^^^^^^
  |          |         |
  |          |         `Option` is not local
  |          `Trait` is not local
  |
  = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
  = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
help: move this `impl` block outside of the current function `main`
 --> a.rs:6:5
  |
6 |     impl Trait for Option<Local> {}
  |     ^^^^^^^^^^^^^^^^^^^^^^-----^^^^
  |                           |
  |                           may need to be moved as well
  = note: this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
  = note: `#[warn(non_local_definitions)]` on by default
```

<details>
<summary>with colors</summary>

![image](https://github.com/rust-lang/rust/assets/3616612/b782a199-650b-4d38-af6b-ebd41d1984cd)


</details>

estebank commented on tests/ui/lint/non-local-defs/weird-exprs.stderr:
```diff
@@ -1,71 +1,111 @@
+   |
+   = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
+   = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
+help: move this `impl` block outside of the current constant expression `<unnameable>`
```
This is a minor gripe, but you should have one message for when the expression is nameable, and one for when it isn't to avoid this.

Edit: it actually seems like every case we show this help it is `<unnameable>`, so we can just remove that part of the message.

estebank replied:
Instead of this help, why don't we make it a secondary span on the main diagnostic stating something along the lines of "move the `impl` block outside of this ...":

```

warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
  --> $DIR/weird-exprs.rs:8:5
   |
LL |/ type A = [u32; {
LL ||     impl Uto for *mut Test {}
   ||     ^^^^^---^^^^^---------
   ||          |       |
   ||          |       `*mut Test` is not local
   ||          `Uto` is not local
LL ||     1
LL || }];
   ||___- move the `impl` block outside of this `type`
   |
   = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
   = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
   = note: this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
   = note: `#[warn(non_local_definitions)]` on by default
```

Urgau replied:
I like the idea, but for how would you expect it to look like for the maximal case (ie where we have multiple suggestions?), putting everything in the main span looks like this:

```text
warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
  --> $DIR/exhaustive.rs:58:5
   |
LL | fn main() {
   | --------- move this `impl` block outside of the current function `main`
...
LL |     impl Trait for *mut InsideMain {}
   |     ^^^^^-----^^^^^---------------
   |          |         |    |
   |          |         |    may need to be moved as well
   |          |         `*mut InsideMain` is not local
   |          |         help: remove `*mut ` to make the `impl` local
   |          `Trait` is not local
   |
   = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
   = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
```
this looks unreadable, any idea how to improve it?

Urgau replied:
Maybe like this, where the "may need to be moved" spans point to the definition type?
```text
warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
  --> /home/archie/Projects/RustProjects/rust/tests/ui/lint/non-local-defs/exhaustive.rs:58:5
   |
9  | fn main() {
   | --------- move this `impl` block outside of the current function `main`
...
56 |     struct InsideMain;
   |     ----------------- may need to be moved as well
57 |
58 |     impl Trait for *mut InsideMain {}
   |     ^^^^^-----^^^^^---------------
   |          |         |
   |          |         `*mut InsideMain` is not local
   |          |         help: remove `*mut ` to make the `impl` local
   |          `Trait` is not local
   |
   = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
   = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
   = note: this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
```

Urgau replied:
But that can quickly get out of hand.
```text
warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
  --> $DIR/suggest-moving-inner.rs:12:5
   |
LL | fn main() {
   | --------- move this `impl` block outside of the current function `main`
LL |     mod below {
LL |         pub struct Type<T>(T);
   |         ------------------ may need to be moved as well
LL |     }
LL |     struct InsideMain;
   |     ----------------- may need to be moved as well
LL |     trait HasFoo {}
   |     ------------ may need to be moved as well
LL |
LL |     impl<T> Trait<InsideMain> for &Vec<below::Type<(InsideMain, T)>>
   |     ^^^^^^^^-----^^^^^^^^^^^^^^^^^----------------------------------
   |             |                     |
   |             |                     `&'_ Vec<below::Type<(InsideMain, T)>>` is not local
   |             `Trait` is not local
   |
   = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
   = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
```

Urgau replied:
Or maybe a mix of both?
```text
warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
  --> /home/archie/Projects/RustProjects/rust/tests/ui/lint/non-local-defs/suggest-moving-inner.rs:12:5
   |
12 |     impl<T> Trait<InsideMain> for &Vec<below::Type<(InsideMain, T)>>
   |     ^^^^^^^^-----^^^^^^^^^^^^^^^^^----------------------------------
   |             |                     |
   |             |                     `&'_ Vec<below::Type<(InsideMain, T)>>` is not local
   |             `Trait` is not local
   |
   = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
   = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
help: move this `impl` block outside of the current function `main`
  --> /home/archie/Projects/RustProjects/rust/tests/ui/lint/non-local-defs/suggest-moving-inner.rs:5:1
   |
5  | fn main() {
   | ^^^^^^^^^
6  |     mod below {
7  |         pub struct Type<T>(T);
   |         ------------------ may need to be moved as well
8  |     }
9  |     struct InsideMain;
   |     ----------------- may need to be moved as well
10 |     trait HasFoo {}
   |     ------------ may need to be moved as well
   = note: this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
   = note: `#[warn(non_local_definitions)]` on by default
```

Urgau replied:
Okay after some experiment which you can see above, I pushed a mix between your proposition and my last one (for more complicated cases), I think it strikes a good balance, yours becomes the default in the vast majority of cases. Let me known what you think.

estebank commented on tests/ui/lint/non-local-defs/weird-exprs.stderr:
```diff
@@ -1,71 +1,111 @@
-   = help: move this `impl` block outside the of the current constant expression `<unnameable>`
-   = note: an `impl` definition is non-local if it is nested inside an item and may impact type checking outside of that item. This can be the case if neither the trait or the self type are at the same nesting level as the `impl`
-   = note: one exception to the rule are anon-const (`const _: () = { ... }`) at top-level module and anon-const at the same nesting as the trait or type
+LL |     impl Uto for *mut Test {}
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
```
This span seems redundant to me. Maybe if you point to where it *should* live (which we can do as a separate PR), but as is it seems better to make this a note instead of a span help.

For reference, this is roughly what would be required to find an appropriate span:

https://github.com/rust-lang/rust/blob/1a7397988684934ae01a71f524bdfff24895d8cc/compiler/rustc_borrowck/src/diagnostics/move_errors.rs#L371-L399

estebank commented on tests/ui/lint/non-local-defs/suggest-moving-inner.stderr:
```diff
@@ -0,0 +1,31 @@
+help: move this `impl` block outside of the current function `main`
+  --> $DIR/suggest-moving-inner.rs:12:5
+   |
+LL |       impl<T> Trait<InsideMain> for &Vec<below::Type<(InsideMain, T)>>
+   |       ^             ----------           -----------  ---------- may need to be moved as well
```
I see that you're hedging your bets that the type itself might also be in a weird scope that would otherwise not be accessible to the `impl` in the new location. Ideally we would check that that is the case before saying anything.

Urgau replied:
I don't understand your comment, it's not betting, we are checking in the code if the type is at the same nesting, and if it is we emit the note saying that it might (as in may not be something you might want to do) need to be moved, there is no guess here.

estebank replied:
Ah! I see. In that case I would make it a message explaining alternative ways of sealing traits and types?

Urgau replied:
It doesn't make sense to recommend sealing the trait since if we warn that means it already not local so moving the `impl` shouldn't change it's availability/visibility.

Regarding sealing the type, idk if that's a thing, but it also wouldn't change anything, since if it is not local it is already leaked.

In general it is very difficult to say how users should change their code to make a impl local (which from the type-system point-of-view doesn't exist anyway), so in general the lint "fallback" to proposing to make the impl def true to what it is, a global impl, and as such should be in a "global context" (module) and not inside a function, consts, array count, ...

estebank commented on compiler/rustc_lint/src/non_local_def.rs:
```diff
@@ -215,17 +191,77 @@ impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
+                let impl_span = item.span.shrink_to_lo().to(impl_.self_ty.span);
+                let mut ms = MultiSpan::from_span(impl_span);
+
+                let (self_ty_span, self_ty_str) =
+                    self_ty_kind_for_diagnostic(&impl_.self_ty, cx.tcx);
```
Given that this is only used to get the type name for the type and trait in the `impl` block, I would avoid it: we're already pointing at the appropriate spans, so naming them is redundant with just the line above. Instead you can just leave it as "this is not local" and that's it. The only case where naming the type is useful is when it ends up in a message of a note or help, on its own and will otherwise not be seen in context (which can also happen with rust-analyzer).

This change would also preclude the need of depending on hir_pretty, which is an added bonus.

Urgau replied:
I wanted to mentioned them for the macro generated case since otherwise the span is just pointing to the macro it-self which isn't helpful.

estebank commented on tests/ui/lint/non-local-defs/cargo-update.stderr:
```diff
@@ -1,14 +1,22 @@
-warning: non-local `impl` definition, they should be avoided as they go against expectation
+warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
   --> $DIR/cargo-update.rs:17:1
    |
 LL | non_local_macro::non_local_impl!(LocalStruct);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   | |
+   | `LocalStruct` is not local
+   | `Debug` is not local
```
We should customize the error when involving macros, because the suggestion to move the impl out of a const is not actionable for users in this case. Even further, it is likely that we shouldn't emit the lint if it comes from a foreign macro (do not deal with that in this PR).

Urgau replied:
I agree it would be good if we could have a specialized version of the diagnostic for macro-generated code, but I have no idea what it would look like.

> Even further, it is likely that we shouldn't emit the lint if it comes from a foreign macro (do not deal with that in this PR).

We must still emit the warning in macro-generated code, since the goal is to make the lint deny-by-default and then hard-error in a future edition, and without the warn it would go straight to hard-error which is not good.

estebank replied:
As follow up work (not this PR), we should still customize the error to state "the macro needs to be changed".

estebank reviewed (changes requested):
There's quite a bit of code that I need to continue reviewing, but I have some nitpicks based on the output.

Urgau committed:
non_local_defs: point the parent item when appropriate

Urgau commented:
All the review comments have been addressed or have been replied to.

@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #124417) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Urgau commented:
@estebank could you give this PR another review.

estebank reviewed (approved):
Minor nitpick in wording, beyond that this is ready to land.

Urgau commented on compiler/rustc_lint/messages.ftl:
```diff
@@ -542,17 +542,21 @@ lint_non_local_definitions_cargo_update = the {$macro_kind} `{$macro_name}` may
 
-lint_non_local_definitions_macro_rules = non-local `macro_rules!` definition, they should be avoided as they go against expectation
+lint_non_local_definitions_impl_move_help =
+    move this `impl` block outside of the current {$body_kind_descr} {$depth ->
```
Changed.

estebank commented:
@bors r+

bors commented:
:pushpin: Commit c7d300442ff94cbe60a22400750cd03dbf15bcef has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank c7d300442ff94cbe60a22400750cd03dbf15bcef -->
<!-- homu: {"type":"Approved","sha":"c7d300442ff94cbe60a22400750cd03dbf15bcef","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
